[Messenger] Add Transport attribute to configure message to transport mapping

maxim-dovydenok opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes<!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | Fix #41106 <!-- prefix each issue number with "Fix #", no need to create an issue if none exist, explain below instead -->
| License       | MIT
| Doc PR        | symfony/symfony-docs#15332 <!-- required for new features -->

This PR adds `Transport` attribute to configure Message->Transport mapping. This attribute is read in the SenderLocator class.
Using this attribute it will be possible to configure transport for the message in the message class file instead of messenger.yaml.

<!--
Replace this notice by a short README for your feature/bugfix. This will help people
understand your PR and can be used as a start for the documentation.

Additionally (see https://symfony.com/releases):
 - Always add tests and ensure they pass.
 - Never break backward compatibility (see https://symfony.com/bc).
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against branch 5.x.
 - Changelog entry should follow https://symfony.com/doc/current/contributing/code/conventions.html#writing-a-changelog-entry
-->

maxim-dovydenok committed:
[Messenger] Add Senders attribute to configure message to sender mapping

maxim-dovydenok committed:
[Messenger] Enable attribute reading only for PHP 8

maxim-dovydenok committed:
[Messenger] Add entry to changelog

maxim-dovydenok committed:
[Messenger] Add license header and fix code style

maxim-dovydenok committed:
[Messenger] Remove strict types

ro0NL commented on src/Symfony/Component/Messenger/Attribute/Senders.php:
```diff
@@ -0,0 +1,29 @@
+ * @author Maxim Dovydenok <dovydenok.maxim@gmail.com>
+ */
+#[\Attribute(\Attribute::TARGET_CLASS)]
+class Senders
```
AFIAK we get the sender from a transport, so perhaps `#[Transport(name)]`?

also i tend to prefer repeatable singulars, for more granularity and potential better composition

ro0NL commented on src/Symfony/Component/Messenger/Transport/Sender/SendersLocator.php:
```diff
@@ -41,20 +42,52 @@ public function __construct(array $sendersMap, ContainerInterface $sendersLocato
+        $senderAliases = [];
+
+        if (\PHP_VERSION_ID >= 80000) {
+            $senderAliases = $this->getSendersFromAttributes($envelope);
```
should semantic config merge or overwrite? (i think overwrite for now)

maxim-dovydenok replied:
I merged them to keep transports from parent classes and interfaces. I'm not sure it will be obvious that the attribute will cancel the routing configured for its interface.

BTW, since I changed Senders to Transport - we don't have an option to configure behaviour with an argument on an attribute (because we can't have different behaviour for different transports of one class), but it can be done using another attribute (probably an overkill).

ro0NL commented on src/Symfony/Component/Messenger/Transport/Sender/SendersLocator.php:
```diff
@@ -41,20 +42,52 @@ public function __construct(array $sendersMap, ContainerInterface $sendersLocato
-                    $sender = $this->sendersLocator->get($senderAlias);
-                    yield $senderAlias => $sender;
-                }
+                $senderAliases[] = $senderAlias;
```
~previous FIFO, now LIFO~ oh i think array_unique preserves it

maxim-dovydenok replied:
Wow, I didn't even think about it. Yes, array_unique should keep the same order.

maxim-dovydenok committed:
[Messenger] Fix code review comments

Overwrite config routing
Read from parent interfaces and classes
Add tests

maxim-dovydenok commented:
Added some changes:
- renamed Senders to Transport mapping and made it repeatable
- made it overwrite file config
- added reading from parent classes and interfaces
- and added tests for the cases above

maxim-dovydenok committed:
[Messenger] Added test for multiple attributes

ro0NL commented:
> made it overwrite file config

sorry, i meant file config should overwrite attributes. So we have a way to opt-out from whatever the code is telling us.

without attributes AND file config, we could maybe inherit the parent attributes (as we inherit file config). Meaning we also have a way to opt out from the parent attr.

OskarStark commented on src/Symfony/Component/Messenger/CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@ CHANGELOG
 5.3
 ---
 
+ * Add `Transport` attribute to configure message to transport mapping
```
For 5.4 please

maxim-dovydenok commented:
> sorry, i meant file config should overwrite attributes. So we have a way to opt-out from whatever the code is telling us.

So lets say we have:
```
#[Transport('from_attribute')]
class Message implements MessageInterface
// and
routing:
  App/Message/MessageInterface: 'from_interface_config'
  App/Message/Message: 'from_config'
```
Do you expect it to be sent to `from_interface_config`, `from_config`? - attributes are ignored since config is present.

> without attributes AND file config, we could maybe inherit the parent attributes (as we inherit file config). Meaning we also have a way to opt out from the parent attr.

Why would we want to opt-out from parent attributes? I mean we don't have an option to opt-out from MessageInterface routing (see example above) currently, at least that I'm aware of, do we?

maxim-dovydenok committed:
[Messenger] Overwrite attributes if framework's config is specified

ro0NL commented:
> Do you expect it to be sent to from_interface_config, from_config? - attributes are ignored since config is present

Yes.

I see i made a thinking error. Each type either derives from file config, or if unavailable, attributes. Meaning we can opt-out from the parent by also specifying file config for that type.

nicolas-grekas commented:
What's the way to configure a transport right now using a DI tag?
The implementation should reuse that tag instead of harcoding the attribute deep in SenderLocator.

maxim-dovydenok commented:
> What's the way to configure a transport right now using a DI tag?
> The implementation should reuse that tag instead of harcoding the attribute deep in SenderLocator.

At the moment we have this code in FrameworkExtension. It reads routing from the config and passes the message-to-senders map to the SendersLocator (where I did changes).
```php
        $messageToSendersMapping = [];
        foreach ($config['routing'] as $message => $messageConfiguration) {
            if ('*' !== $message && !class_exists($message) && !interface_exists($message, false)) {
                throw new LogicException(sprintf('Invalid Messenger routing configuration: class or interface "%s" not found.', $message));
            }

            foreach ($messageConfiguration['senders'] as $sender) {
                if (!isset($senderReferences[$sender])) {
                    throw new LogicException(sprintf('Invalid Messenger routing configuration: the "%s" class is being routed to a sender called "%s". This is not a valid transport or service id.', $message, $sender));
                }
            }

            $messageToSendersMapping[$message] = $messageConfiguration['senders'];
        }

        $sendersServiceLocator = ServiceLocatorTagPass::register($container, $senderReferences);

        $container->getDefinition('messenger.senders_locator')
            ->replaceArgument(0, $messageToSendersMapping)
            ->replaceArgument(1, $sendersServiceLocator)
        ;
```
There's also code, that uses `$config['transports']` to add transports to the container. Each transport has a `messenger.receiver` tag, but I'm not sure how it can help us to configure routing using attributes.

ro0NL commented on src/Symfony/Component/Messenger/Tests/Transport/Sender/SendersLocatorTest.php:
```diff
@@ -15,12 +15,47 @@
+        );
+
+        $locatorWithRouting = new SendersLocator([
+            DummyMessageWithAttribute::class => ['my_merged_sender'],
```
i think we need a test for `DummyMessageWithAttributeAndInterface` as well ;)

ro0NL reviewed (commented):
@nicolas-grekas this is indeed about [routing-messages-to-a-transport](https://symfony.com/doc/current/messenger.html#routing-messages-to-a-transport)

my main motivation here is, to have a single list of things; the messages in code. Or reduce config management if you like :)

maxim-dovydenok committed:
[Messenger] Add tests and rename senders

maxim-dovydenok committed:
[Messenger] Change overwriting behavior, overwrite per type

maxim-dovydenok committed:
[Messenger] Fix tests on 7.2

alsar commented:
@shiftby Is this still planed for 5.4?

kbond commented on src/Symfony/Component/Messenger/Attribute/Transport.php:
```diff
@@ -0,0 +1,26 @@
+#[\Attribute(\Attribute::TARGET_CLASS | \Attribute::IS_REPEATABLE)]
+class Transport
+{
+    public string $name;
+
+    public function __construct(string $name)
+    {
+        $this->name = $name;
+    }
```
```suggestion
    public function __construct(public string $name)
    {
    }
```

kbond commented on src/Symfony/Component/Messenger/Transport/Sender/SendersLocator.php:
```diff
@@ -41,20 +42,56 @@ public function __construct(array $sendersMap, ContainerInterface $sendersLocato
+                $typeSenderAliases[] = $senderAlias;
+            }
+
+            if (\PHP_VERSION_ID >= 80000 && empty($typeSenderAliases)) {
```
Can remove php version checks.

kbond commented on src/Symfony/Component/Messenger/Tests/Transport/Sender/SendersLocatorTest.php:
```diff
@@ -15,12 +15,75 @@
 class SendersLocatorTest extends TestCase
 {
+    /**
+     * @requires PHP 8
```
Can remove.

kbond reviewed (dismissed):
I really like this. I think a common pattern is for apps to create `Sync`/`Async` marker interfaces to simplify routing rules.

kbond committed:
cleanup

kbond commented:
@shiftby, I took the liberty of rebasing and cleaning this up - hope we can get this in 6.1.

@ro0NL, @nicolas-grekas, could we get a fresh review?

weaverryan commented on src/Symfony/Component/Messenger/Transport/Sender/SendersLocator.php:
```diff
@@ -41,20 +42,56 @@ public function __construct(array $sendersMap, ContainerInterface $sendersLocato
+                $typeSenderAliases[] = $senderAlias;
+            }
+
+            if (empty($typeSenderAliases)) {
```
```suggestion
            if (0 === \count($typeSenderAliases)) {
```

I believe is what we prefer over empty

weaverryan commented on src/Symfony/Component/Messenger/Transport/Sender/SendersLocator.php:
```diff
@@ -41,20 +42,56 @@ public function __construct(array $sendersMap, ContainerInterface $sendersLocato
+                $typeSenderAliases = $this->getSendersFromAttributes($type);
+            }
+
+            $senderAliases = array_merge($senderAliases, $typeSenderAliases);
```
If I understand it correctly, IF even one *real* route exists for the message (something in `$this->sendersMap[$type]`), then we do NOT also add the `#[Transport]` senders? So basically any config would *override* the attribute?

I think that is... probably the correct behavior. But what about the `*` catch-all route in `messenger.yaml` (or an interface)? I'm wondering if someone will have the use-case where they generally use the attribute, but then they decide to, in addition to the transport in each class, route all messages (or all messages for an interface) to a 2nd transport via the routing config in `messenger.yaml`. 

So, I'm conflicted on this point.

weaverryan commented on src/Symfony/Component/Messenger/Transport/Sender/SendersLocator.php:
```diff
@@ -41,20 +42,56 @@ public function __construct(array $sendersMap, ContainerInterface $sendersLocato
+            $attributeInstance = $attribute->newInstance();
+
+            return $attributeInstance->name;
+        }, $attributes);
```
It's unfortunate to need to do this at runtime (and not compile time). But as messages are not services... or tagged in any way, there is no way for us to collect the attribute information at container build time. So 👍 for this approach.

kbond replied:
This is indeed unfortunate. Any clever ideas? I understand _it depends_, but is there a non-negligible performance impact to processing the attributes at runtime?

weaverryan reviewed (commented):
I would totally use this :)

kbond committed:
fix per review

kbond committed:
allow subclasses of `Transport`

kbond commented:
I made a little change to allow sub-classes of the `Transport` attribute. This allows you to create custom, app-specific attributes to attach to your messages:

```php
#[\Attribute(\Attribute::TARGET_CLASS)]
class HighPriority extends Transport
{
    public function __construct()
    {
        parent::__construct('high_priority');
    }
}

#[HighPriority]
class MyMessage {}
```

ro0NL commented:
we need additional transport metadata per messsage, so im working on a userland solution instead

that said, the route i consider is `#[WithTransportMetadata(names: ['async'], something: false)]`

ideally i'd like the mapping resolution happening at compile-time as well

this can be done by obtaining the message class from the handler service

ro0NL commented:
turns out it was this simple: https://github.com/etrias-nl/php-toolkit/blob/main/src/Messenger/DependencyInjection/MessengerPass.php

i'd still like to see it in core, put please consider the 'per message transport options' at the abstraction level :angel:

pounard commented:
I'm sorry if I hijacked this, but https://github.com/symfony/symfony/pull/57507 is ready. This PR is stalled for approximatly a year, and it will cause conflict if merged over the most recent Symfony version, this would mean rewrite pretty much everything.

Should we close this PR?

nicolas-grekas commented:
Let us know if #57507 isn't enough. Thanks @maxim-dovydenok for pushing this forward.
